planner/core/expression_rewriter.go

```diff
@@ -400,6 +400,13 @@ func (er *expressionRewriter) buildSemiApplyFromEqualSubq(np LogicalPlan, l, r e
 		if lCol, ok := l.(*expression.Column); ok && mysql.HasNotNullFlag(lCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
 			rCol.InOperand = false
 		}
+		if lCon, ok := l.(*expression.Constant); ok && !lCon.Value.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
+			rCol.InOperand = false
+		}
+		lClc, ok := l.(*expression.CorrelatedColumn)
+		if ok && !lClc.Data.IsNull() && mysql.HasNotNullFlag(rCol.GetType().Flag) {
```

Comment:
`Data` is filled in execution, it has not been initialized here. I think we should check whether `Column` of correlated column has NotNull flag instead.

How about:
```golang
corCol, ok := l.(*expression.CorrelatedColumn)
if ok && mysql.HasNotNullFlag(corCol.GetType().Flag) && mysql.HasNotNullFlag(rCol.GetType().Flag) {
```